### ISSUE TITLE
bug: `HttpRequestError`, `AbiFunctionSignatureNotFoundError` and `TimeoutError` when launching `yarn chain` as a forked endivonment

### ISSUE THREAD

**Issue by luksgrin** [scaffold-eth/scaffold-eth-2#524]
### Is there an existing issue for this?

- [X] I have looked through the [existing issues](https://github.com/scaffold-eth/scaffold-eth-2/issues)

### Current Behavior

When executing a local anvil as a forked environment, the blockexplorer page raises the following error:

```
Cannot connect to local provider
- Did you forget to run yarn chain ?
- Or you can change targetNetwork in scaffold.config.ts
```

The `chainId` remains 31337 even if it's a forked environment (just like when running `yarn chain`). After further inspection, my browser's console also throws (many consecutive times until the app freezes and crashes) the following exceptions:

- `AbiFunctionSignatureNotFoundError `

```
Parsing failed: AbiFunctionSignatureNotFoundError: Encoded function signature "0xe797ec1b" not found on ABI.
Make sure you are using the correct ABI and that the function exists on it.
You can look up the signature here: https://openchain.xyz/signatures?query=0xe797ec1b.

Docs: https://viem.sh/docs/contract/decodeFunctionData.html
Version: viem@1.2.1
window.console.error @ next-dev.js:20
decodeTransactionData @ decodeTxData.ts:38
eval @ useFetchBlocks.ts:46
eval @ useFetchBlocks.ts:46
eval @ useFetchBlocks.ts:45
await in eval (async)
eval @ useFetchBlocks.ts:72
commitHookEffectListMount @ react-dom.development.js:23150
commitPassiveMountOnFiber @ react-dom.development.js:24926
commitPassiveMountEffects_complete @ react-dom.development.js:24891
commitPassiveMountEffects_begin @ react-dom.development.js:24878
commitPassiveMountEffects @ react-dom.development.js:24866
flushPassiveEffectsImpl @ react-dom.development.js:27039
flushPassiveEffects @ react-dom.development.js:26984
eval @ react-dom.development.js:26769
workLoop @ scheduler.development.js:266
flushWork @ scheduler.development.js:239
performWorkUntilDeadline @ scheduler.development.js:533
next-dev.js
```

- `TimeoutError`

```
TimeoutError: The request took too long to respond.

URL: http://127.0.0.1:8545
Request body: {"method":"eth_getBlockByNumber","params":["0x933d44",true]}

Details: The request timed out.
Version: viem@1.2.1
    at Object.http (rpc.js:28:1)
    at fn (http.js:39:30)
    at request (http.js:40:41)
    at delay.count.count (buildRequest.js:26:1)
    at attemptRetry (withRetry.js:12:1)
    at eval (withRetry.js:22:1)
    at new Promise (<anonymous>)
    at withRetry (withRetry.js:3:1)
    at Object.eval [as request] (buildRequest.js:24:38)
    at fetch (fallback.js:16:1)
    at request (fallback.js:48:1)
    at delay.count.count (buildRequest.js:26:1)
    at attemptRetry (withRetry.js:12:1)
    at retry (withRetry.js:9:1)	
```

- `HttpRequestError`

```
URL: http://127.0.0.1:8545
Request body: {"method":"eth_getTransactionReceipt","params":["0x047bd5d6dca285e48878e8f6b85f95361c0c5a18bdb323dc24d99bf9dd6bb598"]}

Details: Failed to fetch
Version: viem@1.2.1
    at Object.http (rpc.js:55:1)
    at async fn (http.js:39:1)
    at async request (http.js:40:41)
    at async delay.count.count (buildRequest.js:26:1)
    at async attemptRetry (withRetry.js:12:1)
```

On a side note, if I kill the app (and chain) before it completely freezes, it ends up showing the 20 most recent transactions, which is part of the normal behavior.

### Expected Behavior

The expected behavior is that the blockexplorer page shows the latest 20 transactions while no errors are being thrown nor freezes/crashes occurred, even if the local chain results from a forked environment

### Steps To Reproduce

Install a scaffold-eth project:

```bash
npx create-eth@latest
```

choose `foundry` as the backend and install all packages.

Next, define a local `.env` file where we are storing the rpc url used to fork. I am forking goerli so:

```bash
echo "GOERLI_RPC_URL=<rpc-url-with-key>" > .env
```

Load this environmental variable

```bash
source .env
```

launch the forked chain

```bash
yarn chain --fork-url $GOERLI_RPC_URL
```

and launch the frontend on another terminal

```bash
yarn start
```

Finally open `localhost:3000/blockexplorer` on the browser.

_(performing all these steps but without the `--fork-url $GOERLI_RPC_URL` parameter results in a working app, just as intended)_

### Anything else?

# Screenshots

During the crash, it looks like this:

![image](https://github.com/scaffold-eth/scaffold-eth-2/assets/49839840/8789adee-b291-456a-9158-989bc8c07ff6)

After killing the app in the terminal, it loads everything into the frontend:

![image](https://github.com/scaffold-eth/scaffold-eth-2/assets/49839840/d363229f-ed8f-4628-97c2-2cf38643206e)

# System Information

```
uname -a:
Linux be79f216f043 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

```
yarn --version:
3.2.3
```

**Comment by luksgrin** [comment]
### Why are you doing this though?

I have a foundry project which I heavily tested locally through a local fork of the Goerli testnet (because my contracts interact with deployed contracts such as WETH and AAVE's lendingPool contract).

Now I've come to the point of building a front-end for my dApp, and since scaffold-eth supports foundry, I created a scaffold-eth project and migrated my whole foundry project into it (adapting the deployment scripts accordingly etc...), and then I started encountering this bug.

At first, I thought that the issue lied in my migration, but I ended up hunting down the bug to the forked chain (as you can see in the above issue, I haven't modified/deployed a smart contract at all).

**Comment by luksgrin** [comment]
### UPDATE

It seems that, for some reason, these exceptions are all thrown by the `useFetchBlocks()` hook.

**Comment by technophile-04** [comment]
Hey @luksgrin thanks for raising the issue and detailed explanation, really appreciate it !! 

It seems that when you fork a chain the chain id is changed in this case to `5` but you need to point to localhost RPC url 

Because of this inconsistency and the way how Scaffold-ETH2 is configured, it fails to work nicely 

I tried tinkering with it a bit, you can have look at this PR and make changes to your own repo : 
https://github.com/technophile-04/foundry-goerli-fork-524/pull/1/files

https://github.com/scaffold-eth/scaffold-eth-2/assets/80153681/fb0774b7-9427-4282-b76b-19759be4e24c

NOTE: After loading is complete, there is slight blank screen with no data but its internally fetching all transactionReceipt so you have to wait a bit to see all the data (This needs to be handle)

Also, few things to note : 

1. The logic for blockexplorer is naive and not optimised (which is fine since it was written keeping in mind to power local chain)
2. So the way logic works : It iteratively fetch each block data and also iteratively travels through each Txn Recipet for each block
3. Since we have forked the chain, by using blockexplorer page you are gonna exhaust your Alchemy RPC by making lot of `eth_getBlockByIndex` , `eth_getTransactionReceipt`

Nevertheless, this issue is free to grab if anyone wants to tackle it 🙌

**Comment by luksgrin** [comment]
Thank you @technophile-04 for your kind answer. I was able to make the blockexplorer work thanks to the modifications you did on [your PR](https://github.com/technophile-04/foundry-goerli-fork-524/pull/1/files) . Nonetheless, it perplexes me that while loading, I still get errors:

![image](https://github.com/scaffold-eth/scaffold-eth-2/assets/49839840/de6a2daf-037c-44ad-af53-37115ebfe41b)

I believe these errors are silent due to the [removal of the error return](https://github.com/technophile-04/foundry-goerli-fork-524/pull/1/files#diff-4fd315ccb3cf006c1a2cad112dc57f572ce7e51c8450b60ff890280582652a4dR50).

Also, would it be a way to bypass manually writing the settings for the GoerliFork? Initially I believed I could bypass the chainId issue by launching anvil as:

```bash
yarn chain --fork-url $GOERLI_RPC_URL --chain-id 31337
```
this way I would force it to continue being `chainId 31337` instead of Goerli's `chainId 5`.

![image](https://github.com/scaffold-eth/scaffold-eth-2/assets/49839840/00abbffd-af22-4788-8108-2f452d5a436f)

But this didn't help...

I will keep messing around with it. Maybe I can find something out. I would truly appreciate further comments/hints 😄

**Comment by luksgrin** [comment]
## UPDATE

So, I wanted to tackle first the apparent issue of the `chainId`. Indeed, one can control its value with the `--chain-id` flag. Because executing

```bash
yarn chain --fork-url $GOERLI_RPC_URL --chain-id 31337
```

and then calling

```bash
curl http://localhost:8545 \
-X POST \
-H "Content-Type: application/json" \
-d '{"jsonrpc":"2.0","method":"eth_chainId","params": [],"id":1}'
```

immediately returned

```bash
{"jsonrpc":"2.0","id":1,"result":"0x7a69"}
```

_(being `0x7a69` the hex representation of `31337`)_. Given that within `wagmi/chains` (which calls `viem/chains`) the settings for `hardhat` and `Goerli` only changed on the `chainId` value, I reverted all modifications [on the PR](https://github.com/technophile-04/foundry-goerli-fork-524/pull/1/files) **except the [removal of the error return](https://github.com/technophile-04/foundry-goerli-fork-524/pull/1/files#diff-4fd315ccb3cf006c1a2cad112dc57f572ce7e51c8450b60ff890280582652a4dR50)**. This way, I reached the same results as previously, i.e.

```bash
POST http://127.0.0.1:8545/ net::ERR_INSUFFICIENT_RESOURCES
```

which after freezing for a bit, renders a fully functional block explorer. I do believe the issue may be in the way transactions are queried [as mentioned here](https://github.com/scaffold-eth/scaffold-eth-2/issues/524#issuecomment-1711082748). I will try to optimize it so that my app does not crash under my specific conditions. I believe we are close! 😄

**Comment by luksgrin** [comment]
## FINAL UPDATE

<details>
<summary><b>The long story</b></summary>
I did not optimize the transaction querying. To retrieve massive data, providers such as Alchemy should be preferred. Nonetheless, <a href=https://github.com/technophile-04/foundry-goerli-fork-524/blob/foundry-goerli-fork/packages/nextjs/hooks/scaffold-eth/useFetchBlocks.ts#L7>I found that reducing the number of blocks to query (from 20 to 10)</a> substantially lowers the requests (duh!) and my app does not die. The errors are still present, but at least it renders in a reasonable amount of time. There's still <a href=https://github.com/technophile-04/foundry-goerli-fork-524/pull/1/files#diff-4fd315ccb3cf006c1a2cad112dc57f572ce7e51c8450b60ff890280582652a4dR50>the option of removing the error return so that the front end is not polluted<a>, you know, for a clean demo video 😉 

I am satisfied with this fix, and will remember it in case I find myself in a similar scenario in the future.
</details>

Thank you so much @technophile-04 for your response. 

**TLDR;**
The cause of this issue is that the [`useFetchBlocks.ts`](https://github.com/technophile-04/foundry-goerli-fork-524/blob/foundry-goerli-fork/packages/nextjs/hooks/scaffold-eth/useFetchBlocks.ts) is not optimized, so when forking a blockchain, it performs a huge amount of requests that may kill the server.

One can safely fork a blockchain and the app will work properly **as long as the `chainId` is `31337`, which `anvil` allows to manually set via the `--chain-id` setting.

**Comment by technophile-04** [comment]
Tysm @luksgrin for the great research and comments, really appreciate it !!

> So, I wanted to tackle first the apparent issue of the `chainId`. Indeed, one can control its value with the `--chain-id` flag. Because executing 

This was a great finding !! Added it as todo in #470, so that when people use `fork` script from [`pacakge.json`](https://github.com/scaffold-eth/scaffold-eth-2/blob/e979ab45b13af7af69b35ea51cac98f64b68464b/packages/foundry/package.json#L7C2-L7C2) script we by default append `--chaind-id 31337`


Also noticed that this problem does not occur when you choose `hardhat` as solidity framework because forking chain on hardhat keeps the chaind as 31337 🙌 

But thanks again !!

**Comment by carletex** [comment]
Just saying thanks to @luksgrin detailed research/explanation <3

**Comment by technophile-04** [comment]
Closing this for now, I think its recommended that you only use local block explorer when using a simple local chain. 

Using forked chain / visiting local blockexplorer when on chains other than local chain will burn the SE-2 Alchmey RPC / your RPC node if updated  a lot. 

Maybe on SE-2 side we should just return empty page / redirect / show warning message without making any request when you vising blockexplorer and chain is forked( this will be hard to figure out) / other chain.
